### PR TITLE
Advanced Building Blocks - Enumerables (Solo project)

### PR DESCRIPTION
**Pull request by b-a-a-s**
# Implemented custom methods for **Enumerable** module: 
                                                                                                                                                                                                                                                                                                                   
- my_each                                                                                                                                                              
- my_each_with_index                                                                                                                                                   
- my_selec                                                                                                                                                             
- my_all?                                                                                                                                                              
- my_any?                                                                                                                                                              
- my_none?                                                                                                                                                             
- my_count                                                                                                                                                             
- my_map                                                                                                                                                               
- my_inject

### CODE REVIEW

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+module Enumerable
+  def my_each
+    i = 0
+    while i < size
+      yield(self[i])
+      i += 1
+    end
+    self
+  end
```
Comment (maelfosso) [inline enumerables.rb]: Should return an **enumerator**

```ruby
irb(main):006:0> array.each
=> #<Enumerator: [5, 7, 8, 3, 4, 2, 1, 2, 1, 6, 3, 2, 4, 1, 2, 8, 8, 1, 0, 4, 0, 2, 4, 2, 2, 5, 8, 5, 8, 6, 2, 4, 8, 1, 7, 0, 8, 1, 1, 7, 8, 1, 6, 0, 7, 1, 8, 5, 6, 2, 6, 7, 0, 6, 0, 0, 6, 6, 2, 8, 3, 6, 5, 5, 4, 4, 1, 5, 7, 2, 4, 4, 7, 1, 2, 4, 4, 0, 8, 3, 7, 0, 1, 3, 6, 4, 6, 8, 1, 0, 3, 4, 7, 8, 5, 4, 4, 4, 3, 0]:each>
irb(main):024:0> array.my_each
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):24
        1: from enumerables.rb:7:in `my_each'
LocalJumpError (no block given (yield))
irb(main):008:0> array
=> [5, 7, 8, 3, 4, 2, 1, 2, 1, 6, 3, 2, 4, 1, 2, 8, 8, 1, 0, 4, 0, 2, 4, 2, 2, 5, 8, 5, 8, 6, 2, 4, 8, 1, 7, 0, 8, 1, 1, 7, 8, 1, 6, 0, 7, 1, 8, 5, 6, 2, 6, 7, 0, 6, 0, 0, 6, 6, 2, 8, 3, 6, 5, 5, 4, 4, 1, 5, 7, 2, 4, 4, 7, 1, 2, 4, 4, 0, 8, 3, 7, 0, 1, 3, 6, 4, 6, 8, 1, 0, 3, 4, 7, 8, 5, 4, 4, 4, 3, 0]
irb(main):009:0> 
```

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+  def my_each_with_index
+    i = 0
+    while i < size
+      yield(self[i], i)
+      i += 1
+    end
+    self
+  end
```
Comment (maelfosso) [inline enumerables.rb]: Should return an **enumerator**
```ruby
irb(main):011:0> array
=> [5, 7, 8, 3, 4, 2, 1, 2, 1, 6, 3, 2, 4, 1, 2, 8, 8, 1, 0, 4, 0, 2, 4, 2, 2, 5, 8, 5, 8, 6, 2, 4, 8, 1, 7, 0, 8, 1, 1, 7, 8, 1, 6, 0, 7, 1, 8, 5, 6, 2, 6, 7, 0, 6, 0, 0, 6, 6, 2, 8, 3, 6, 5, 5, 4, 4, 1, 5, 7, 2, 4, 4, 7, 1, 2, 4, 4, 0, 8, 3, 7, 0, 1, 3, 6, 4, 6, 8, 1, 0, 3, 4, 7, 8, 5, 4, 4, 4, 3, 0]
irb(main):012:0> array.each_with_index
=> #<Enumerator: [5, 7, 8, 3, 4, 2, 1, 2, 1, 6, 3, 2, 4, 1, 2, 8, 8, 1, 0, 4, 0, 2, 4, 2, 2, 5, 8, 5, 8, 6, 2, 4, 8, 1, 7, 0, 8, 1, 1, 7, 8, 1, 6, 0, 7, 1, 8, 5, 6, 2, 6, 7, 0, 6, 0, 0, 6, 6, 2, 8, 3, 6, 5, 5, 4, 4, 1, 5, 7, 2, 4, 4, 7, 1, 2, 4, 4, 0, 8, 3, 7, 0, 1, 3, 6, 4, 6, 8, 1, 0, 3, 4, 7, 8, 5, 4, 4, 4, 3, 0]:each_with_index>
irb(main):025:0> array.my_each_with_index
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):25
        1: from enumerables.rb:16:in `my_each_with_index'
LocalJumpError (no block given (yield))

```

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+  def my_select
+    arr = []
+    my_each { |e| arr << e if yield(e) }
+    arr
+  end
```
Comment (maelfosso) [inline enumerables.rb]: Should return an **enumerator** if no block given
```ruby
irb(main):021:0> array
=> [5, 7, 8, 3, 4, 2, 1, 2, 1, 6, 3, 2, 4, 1, 2, 8, 8, 1, 0, 4, 0, 2, 4, 2, 2, 5, 8, 5, 8, 6, 2, 4, 8, 1, 7, 0, 8, 1, 1, 7, 8, 1, 6, 0, 7, 1, 8, 5, 6, 2, 6, 7, 0, 6, 0, 0, 6, 6, 2, 8, 3, 6, 5, 5, 4, 4, 1, 5, 7, 2, 4, 4, 7, 1, 2, 4, 4, 0, 8, 3, 7, 0, 1, 3, 6, 4, 6, 8, 1, 0, 3, 4, 7, 8, 5, 4, 4, 4, 3, 0]
irb(main):022:0> array.select
=> #<Enumerator: [5, 7, 8, 3, 4, 2, 1, 2, 1, 6, 3, 2, 4, 1, 2, 8, 8, 1, 0, 4, 0, 2, 4, 2, 2, 5, 8, 5, 8, 6, 2, 4, 8, 1, 7, 0, 8, 1, 1, 7, 8, 1, 6, 0, 7, 1, 8, 5, 6, 2, 6, 7, 0, 6, 0, 0, 6, 6, 2, 8, 3, 6, 5, 5, 4, 4, 1, 5, 7, 2, 4, 4, 7, 1, 2, 4, 4, 0, 8, 3, 7, 0, 1, 3, 6, 4, 6, 8, 1, 0, 3, 4, 7, 8, 5, 4, 4, 4, 3, 0]:select>
irb(main):023:0> array.my_select
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):23
        3: from enumerables.rb:24:in `my_select'
        2: from enumerables.rb:7:in `my_each'
        1: from enumerables.rb:24:in `block in my_select'
LocalJumpError (no block given (yield))
irb(main):024:0> 
```

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+  def my_all?
+    my_each { |e| return false unless yield(e) }
+    true
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```ruby
irb(main):029:0> true_array = [1, true, 'hi', []]
=> [1, true, "hi", []]
irb(main):030:0> true_array.all?
=> true
irb(main):031:0> true_array.my_all?
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):31
        3: from enumerables.rb:29:in `my_all?'
        2: from enumerables.rb:7:in `my_each'
        1: from enumerables.rb:29:in `block in my_all?'
LocalJumpError (no block given (yield))
irb(main):032:0> 
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):034:0> array
=> [5, 7, 8, 3, 4, 2, 1, 2, 1, 6, 3, 2, 4, 1, 2, 8, 8, 1, 0, 4, 0, 2, 4, 2, 2, 5, 8, 5, 8, 6, 2, 4, 8, 1, 7, 0, 8, 1, 1, 7, 8, 1, 6, 0, 7, 1, 8, 5, 6, 2, 6, 7, 0, 6, 0, 0, 6, 6, 2, 8, 3, 6, 5, 5, 4, 4, 1, 5, 7, 2, 4, 4, 7, 1, 2, 4, 4, 0, 8, 3, 7, 0, 1, 3, 6, 4, 6, 8, 1, 0, 3, 4, 7, 8, 5, 4, 4, 4, 3, 0]
irb(main):035:0> array.all?(Integer)
=> true
irb(main):036:0> array.my_all?(Integer)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):36
        1: from enumerables.rb:28:in `my_all?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):037:0> 
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):039:0> words = %w[dog door rod blade]
=> ["dog", "door", "rod", "blade"]
irb(main):040:0> words.all?(/d/)
=> true
irb(main):041:0> words.my_all?(/d/)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):41
        1: from enumerables.rb:28:in `my_all?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):042:0> 
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):044:0> array
=> [5, 7, 8, 3, 4, 2, 1, 2, 1, 6, 3, 2, 4, 1, 2, 8, 8, 1, 0, 4, 0, 2, 4, 2, 2, 5, 8, 5, 8, 6, 2, 4, 8, 1, 7, 0, 8, 1, 1, 7, 8, 1, 6, 0, 7, 1, 8, 5, 6, 2, 6, 7, 0, 6, 0, 0, 6, 6, 2, 8, 3, 6, 5, 5, 4, 4, 1, 5, 7, 2, 4, 4, 7, 1, 2, 4, 4, 0, 8, 3, 7, 0, 1, 3, 6, 4, 6, 8, 1, 0, 3, 4, 7, 8, 5, 4, 4, 4, 3, 0]
irb(main):045:0> array.all?(3)
=> false
irb(main):046:0> array.my_all?(3)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):46
        1: from enumerables.rb:28:in `my_all?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):047:0> 
```

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+  def my_any?
+    my_each { |e| return true if yield(e) }
+    false
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```ruby
irb(main):050:0> true_array.any?
=> true
irb(main):051:0> true_array.my_any?
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):51
        3: from enumerables.rb:34:in `my_any?'
        2: from enumerables.rb:7:in `my_each'
        1: from enumerables.rb:34:in `block in my_any?'
LocalJumpError (no block given (yield))
irb(main):052:0> true_array
=> [1, true, "hi", []]
irb(main):053:0> 
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):055:0> array
=> [5, 7, 8, 3, 4, 2, 1, 2, 1, 6, 3, 2, 4, 1, 2, 8, 8, 1, 0, 4, 0, 2, 4, 2, 2, 5, 8, 5, 8, 6, 2, 4, 8, 1, 7, 0, 8, 1, 1, 7, 8, 1, 6, 0, 7, 1, 8, 5, 6, 2, 6, 7, 0, 6, 0, 0, 6, 6, 2, 8, 3, 6, 5, 5, 4, 4, 1, 5, 7, 2, 4, 4, 7, 1, 2, 4, 4, 0, 8, 3, 7, 0, 1, 3, 6, 4, 6, 8, 1, 0, 3, 4, 7, 8, 5, 4, 4, 4, 3, 0]
irb(main):056:0> array.any?(Integer)
=> true
irb(main):057:0> array.my_any?(Integer)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):57
        1: from enumerables.rb:33:in `my_any?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):058:0> 
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):060:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):061:0> words.any?(/z/)
=> false
irb(main):062:0> words.my_any?(/z/)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):62
        1: from enumerables.rb:33:in `my_any?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):063:0> 
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):065:0> words.any?('cat')
=> false
irb(main):066:0> words.my_any?('cat')
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):66
        1: from enumerables.rb:33:in `my_any?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):067:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):068:0> 
```

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+  def my_none?
+    my_each { |e| return false if yield(e) }
+    true
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```ruby
irb(main):070:0> false_array = [nil, false, nil, false]
=> [nil, false, nil, false]
irb(main):071:0> false_array.none?
=> true
irb(main):072:0> false_array.my_none?
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):72
        3: from enumerables.rb:39:in `my_none?'
        2: from enumerables.rb:7:in `my_each'
        1: from enumerables.rb:39:in `block in my_none?'
LocalJumpError (no block given (yield))
irb(main):073:0> 
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):075:0> array.none?(String)
=> true
irb(main):076:0> array.my_none?(String)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):76
        1: from enumerables.rb:38:in `my_none?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):077:0> array
=> [5, 7, 8, 3, 4, 2, 1, 2, 1, 6, 3, 2, 4, 1, 2, 8, 8, 1, 0, 4, 0, 2, 4, 2, 2, 5, 8, 5, 8, 6, 2, 4, 8, 1, 7, 0, 8, 1, 1, 7, 8, 1, 6, 0, 7, 1, 8, 5, 6, 2, 6, 7, 0, 6, 0, 0, 6, 6, 2, 8, 3, 6, 5, 5, 4, 4, 1, 5, 7, 2, 4, 4, 7, 1, 2, 4, 4, 0, 8, 3, 7, 0, 1, 3, 6, 4, 6, 8, 1, 0, 3, 4, 7, 8, 5, 4, 4, 4, 3, 0]
irb(main):078:0> 
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):080:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):081:0> words.none?(/z/)
=> true
irb(main):082:0> words.my_none?(/z/)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):82
        1: from enumerables.rb:38:in `my_none?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):083:0>
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):085:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):086:0> words.none?(5)
=> true
irb(main):087:0> words.my_none?(5)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):87
        1: from enumerables.rb:38:in `my_none?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):088:0> 
```

File: enumerables.rb [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+  def my_count
+    count = 0
+    my_each { count += 1 }
+    count
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```ruby
irb(main):091:0> array
=> [5, 7, 8, 3, 4, 2, 1, 2, 1, 6, 3, 2, 4, 1, 2, 8, 8, 1, 0, 4, 0, 2, 4, 2, 2, 5, 8, 5, 8, 6, 2, 4, 8, 1, 7, 0, 8, 1, 1, 7, 8, 1, 6, 0, 7, 1, 8, 5, 6, 2, 6, 7, 0, 6, 0, 0, 6, 6, 2, 8, 3, 6, 5, 5, 4, 4, 1, 5, 7, 2, 4, 4, 7, 1, 2, 4, 4, 0, 8, 3, 7, 0, 1, 3, 6, 4, 6, 8, 1, 0, 3, 4, 7, 8, 5, 4, 4, 4, 3, 0]
irb(main):092:0> array.count(0)
=> 11
irb(main):093:0> array.my_count(0)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):93
        1: from enumerables.rb:43:in `my_count'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):094:0> 
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):097:0> block = proc { |num| num < 9 / 2 }
=> #<Proc:0x000055f489654a40@(irb):97>
irb(main):098:0> array.count(&block)
=> 43
irb(main):099:0> array.my_count(&block)
=> 100
irb(main):100:0> 
```

File: enumerables.rb [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+  def my_map(proc)
+    arr = []
+    proc ? my_each { |e| arr << proc.call(e) } : my_each { |e| arr << yield(e) }
+    arr
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```ruby
irb(main):107:0> block = proc { |num| num < 9/2 }
=> #<Proc:0x000055f4894eca90@(irb):107>
irb(main):108:0> array.map(&block)
=> [false, false, false, true, false, true, true, true, true, false, true, true, false, true, true, false, false, true, true, false, true, true, false, true, true, false, false, false, false, false, true, false, false, true, false, true, false, true, true, false, false, true, false, true, false, true, false, false, false, true, false, false, true, false, true, true, false, false, true, false, true, false, false, false, false, false, true, false, false, true, false, false, false, true, true, false, false, true, false, true, false, true, true, true, false, false, false, false, true, true, true, false, false, false, false, false, false, false, true, true]
irb(main):109:0> array.my_map(&block)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):109
        1: from enumerables.rb:49:in `my_map'
ArgumentError (wrong number of arguments (given 0, expected 1))
irb(main):110:0> 
```

File: enumerables.rb [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+  def my_inject(initial = nil)
+    acc = initial || first
+    my_each { |e| acc = yield(acc, e) }
+    acc
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```ruby
irb(main):116:0> operation = proc { |sum, n| sum + n }
=> #<Proc:0x000055f48962ec50@(irb):116>
irb(main):117:0> array
=> [5, 7, 8, 3, 4, 2, 1, 2, 1, 6, 3, 2, 4, 1, 2, 8, 8, 1, 0, 4, 0, 2, 4, 2, 2, 5, 8, 5, 8, 6, 2, 4, 8, 1, 7, 0, 8, 1, 1, 7, 8, 1, 6, 0, 7, 1, 8, 5, 6, 2, 6, 7, 0, 6, 0, 0, 6, 6, 2, 8, 3, 6, 5, 5, 4, 4, 1, 5, 7, 2, 4, 4, 7, 1, 2, 4, 4, 0, 8, 3, 7, 0, 1, 3, 6, 4, 6, 8, 1, 0, 3, 4, 7, 8, 5, 4, 4, 4, 3, 0]
irb(main):118:0> array.inject(&operation)
=> 395
irb(main):119:0> array.my_inject(&operation)
=> 400
irb(main):120:0> 
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):122:0> array
=> [5, 7, 8, 3, 4, 2, 1, 2, 1, 6, 3, 2, 4, 1, 2, 8, 8, 1, 0, 4, 0, 2, 4, 2, 2, 5, 8, 5, 8, 6, 2, 4, 8, 1, 7, 0, 8, 1, 1, 7, 8, 1, 6, 0, 7, 1, 8, 5, 6, 2, 6, 7, 0, 6, 0, 0, 6, 6, 2, 8, 3, 6, 5, 5, 4, 4, 1, 5, 7, 2, 4, 4, 7, 1, 2, 4, 4, 0, 8, 3, 7, 0, 1, 3, 6, 4, 6, 8, 1, 0, 3, 4, 7, 8, 5, 4, 4, 4, 3, 0]
irb(main):123:0> array.inject(:+)
=> 395
irb(main):124:0> array.my_inject(:+)
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):124
        3: from enumerables.rb:57:in `my_inject'
        2: from enumerables.rb:7:in `my_each'
        1: from enumerables.rb:57:in `block in my_inject'
LocalJumpError (no block given (yield))
irb(main):125:0> 
```

File: enumerables.rb (hunk 8 above)
Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):128:0> array
=> [5, 7, 8, 3, 4, 2, 1, 2, 1, 6, 3, 2, 4, 1, 2, 8, 8, 1, 0, 4, 0, 2, 4, 2, 2, 5, 8, 5, 8, 6, 2, 4, 8, 1, 7, 0, 8, 1, 1, 7, 8, 1, 6, 0, 7, 1, 8, 5, 6, 2, 6, 7, 0, 6, 0, 0, 6, 6, 2, 8, 3, 6, 5, 5, 4, 4, 1, 5, 7, 2, 4, 4, 7, 1, 2, 4, 4, 0, 8, 3, 7, 0, 1, 3, 6, 4, 6, 8, 1, 0, 3, 4, 7, 8, 5, 4, 4, 4, 3, 0]
irb(main):129:0> array.map
=> #<Enumerator: [5, 7, 8, 3, 4, 2, 1, 2, 1, 6, 3, 2, 4, 1, 2, 8, 8, 1, 0, 4, 0, 2, 4, 2, 2, 5, 8, 5, 8, 6, 2, 4, 8, 1, 7, 0, 8, 1, 1, 7, 8, 1, 6, 0, 7, 1, 8, 5, 6, 2, 6, 7, 0, 6, 0, 0, 6, 6, 2, 8, 3, 6, 5, 5, 4, 4, 1, 5, 7, 2, 4, 4, 7, 1, 2, 4, 4, 0, 8, 3, 7, 0, 1, 3, 6, 4, 6, 8, 1, 0, 3, 4, 7, 8, 5, 4, 4, 4, 3, 0]:map>
irb(main):130:0> array.my_map
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):130
        1: from enumerables.rb:49:in `my_map'
ArgumentError (wrong number of arguments (given 0, expected 1))
irb(main):131:0> 
```

File: enumerables.rb [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    arr
+  end
+
+  def my_all?(pattern = nil)
+    my_each { |e| return false unless yield(e) } if block_given?
+    my_each { |e| return false unless e.class == pattern } if pattern.class == Class
+    my_each { |e| return false unless e =~ pattern } if pattern.class == Regexp
+    my_each { |e| return false unless e } unless pattern
+    true
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```ruby
irb(main):044:0> array
=> [5, 7, 8, 3, 4, 2, 1, 2, 1, 6, 3, 2, 4, 1, 2, 8, 8, 1, 0, 4, 0, 2, 4, 2, 2, 5, 8, 5, 8, 6, 2, 4, 8, 1, 7, 0, 8, 1, 1, 7, 8, 1, 6, 0, 7, 1, 8, 5, 6, 2, 6, 7, 0, 6, 0, 0, 6, 6, 2, 8, 3, 6, 5, 5, 4, 4, 1, 5, 7, 2, 4, 4, 7, 1, 2, 4, 4, 0, 8, 3, 7, 0, 1, 3, 6, 4, 6, 8, 1, 0, 3, 4, 7, 8, 5, 4, 4, 4, 3, 0]
irb(main):045:0> array.all?(3)
=> false
irb(main):046:0> array.my_all?(3)
=> true
```

File: enumerables.rb [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+  def my_any?(pattern = nil)
+    my_each { |e| return true if yield(e) } if block_given?
+    my_each { |e| return true if e.class == pattern } if pattern.class == Class
+    my_each { |e| return true if e =~ pattern } if pattern.class == Regexp
+    my_each { |e| return true if e } unless pattern
+    false
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```ruby
irb(main):012:0> false_array = proc { |num| num > 9 }
=> #<Proc:0x00005602d062d5a8@(irb):12>
irb(main):013:0> array
=> [0, 3, 8, 3, 6, 8, 1, 3, 0, 7, 5, 6, 0, 3, 6, 7, 6, 4, 5, 6, 4, 3, 7, 6, 2, 7, 2, 0, 2, 2, 0, 8, 2, 4, 0, 8, 8, 7, 3, 2, 8, 0, 4, 6, 6, 5, 8, 0, 6, 0, 7, 5, 8, 4, 3, 5, 3, 4, 6, 5, 3, 6, 5, 1, 8, 3, 4, 7, 4, 6, 7, 0, 7, 1, 3, 1, 3, 5, 2, 5, 4, 2, 0, 7, 8, 2, 0, 0, 2, 4, 7, 5, 4, 0, 8, 1, 4, 3, 0, 7]
irb(main):014:0> array.any?(&false_array)
=> false
irb(main):015:0> array.my_any?(&false_array)
=> true
irb(main):016:0> 
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):021:0> words[0] = 'cat'
=> "cat"
irb(main):022:0> words
=> ["cat", "door", "rod", "blade"]
irb(main):023:0> words.any?('cat')
=> true
irb(main):024:0> words.my_any?('cat')
=> false
irb(main):025:0> 

```

File: enumerables.rb [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+  def my_none?(pattern = nil)
+    my_each { |e| return false if yield(e) } if block_given?
+    my_each { |e| return false if e.class == pattern } if pattern.class == Class
+    my_each { |e| return false if e =~ pattern } if pattern.class == Regexp
+    my_each { |e| return false if e } unless pattern
+    true
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```ruby
irb(main):027:0> true_block = proc { |num| num > 9 }
=> #<Proc:0x00005602d0746fc0@(irb):27>
irb(main):028:0> array
=> [0, 3, 8, 3, 6, 8, 1, 3, 0, 7, 5, 6, 0, 3, 6, 7, 6, 4, 5, 6, 4, 3, 7, 6, 2, 7, 2, 0, 2, 2, 0, 8, 2, 4, 0, 8, 8, 7, 3, 2, 8, 0, 4, 6, 6, 5, 8, 0, 6, 0, 7, 5, 8, 4, 3, 5, 3, 4, 6, 5, 3, 6, 5, 1, 8, 3, 4, 7, 4, 6, 7, 0, 7, 1, 3, 1, 3, 5, 2, 5, 4, 2, 0, 7, 8, 2, 0, 0, 2, 4, 7, 5, 4, 0, 8, 1, 4, 3, 0, 7]
irb(main):029:0> array.none?(&true_block)
=> true
irb(main):030:0> array.my_none?(&true_block)
=> false
irb(main):031:0> 
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):036:0> words = %w[dog door rod blade]
=> ["dog", "door", "rod", "blade"]
irb(main):037:0> words[0] = 5
=> 5
irb(main):038:0> words
=> [5, "door", "rod", "blade"]
irb(main):039:0> words.none?(5)
=> false
irb(main):040:0> words.my_none?(5)
=> true
irb(main):041:0>
```

File: enumerables.rb [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+    my_each { count += 1 } if !item && !block_given?
+    count
+  end
+
+  def my_map(&block)
+    arr = []
+    my_each { |e| arr << block.call(e) } if block_given?
+    arr
+  end
```
Comment (maelfosso) [inline enumerables.rb]: **Should return an enumerator**

```ruby
irb(main):044:0> array
=> [0, 3, 8, 3, 6, 8, 1, 3, 0, 7, 5, 6, 0, 3, 6, 7, 6, 4, 5, 6, 4, 3, 7, 6, 2, 7, 2, 0, 2, 2, 0, 8, 2, 4, 0, 8, 8, 7, 3, 2, 8, 0, 4, 6, 6, 5, 8, 0, 6, 0, 7, 5, 8, 4, 3, 5, 3, 4, 6, 5, 3, 6, 5, 1, 8, 3, 4, 7, 4, 6, 7, 0, 7, 1, 3, 1, 3, 5, 2, 5, 4, 2, 0, 7, 8, 2, 0, 0, 2, 4, 7, 5, 4, 0, 8, 1, 4, 3, 0, 7]
irb(main):045:0> array.map
=> #<Enumerator: [0, 3, 8, 3, 6, 8, 1, 3, 0, 7, 5, 6, 0, 3, 6, 7, 6, 4, 5, 6, 4, 3, 7, 6, 2, 7, 2, 0, 2, 2, 0, 8, 2, 4, 0, 8, 8, 7, 3, 2, 8, 0, 4, 6, 6, 5, 8, 0, 6, 0, 7, 5, 8, 4, 3, 5, 3, 4, 6, 5, 3, 6, 5, 1, 8, 3, 4, 7, 4, 6, 7, 0, 7, 1, 3, 1, 3, 5, 2, 5, 4, 2, 0, 7, 8, 2, 0, 0, 2, 4, 7, 5, 4, 0, 8, 1, 4, 3, 0, 7]:map>
irb(main):046:0> array.my_map
=> []
irb(main):047:0> 
```

File: enumerables.rb [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+  def my_inject(initial = nil, sym = nil, &block)
+    acc = initial || first
+    acc = first if initial.class == Symbol
+    acc = initial if !initial.class == Symbol
+    acc -= acc
+    my_each { |e| acc = block.call(acc, e) } if block_given?
+    my_each { |e| acc = acc.send(sym, e) } if initial && sym
+    my_each { |e| acc = acc.send(initial, e) } if initial.class == Symbol
+    acc
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```ruby
irb(main):051:0> range = Range.new(5, 50)
=> 5..50
irb(main):052:0> range.inject(4) { |prod, n| prod * n }
=> 5069015533618896340602101361010794807396273594826752000000000000
irb(main):053:0> range.my_inject(4) { |prod, n| prod * n }
Traceback (most recent call last):
        4: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        3: from (irb):53
        2: from enumerables.rb:81:in `my_inject'
        1: from enumerables.rb:9:in `my_each'
NoMethodError (undefined method `[]' for 5..50:Range)
irb(main):054:0> 
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):056:0> range
=> 5..50
irb(main):057:0> range.inject(2, :*)
=> 2534507766809448170301050680505397403698136797413376000000000000
irb(main):058:0> range.my_inject(2, :*)
Traceback (most recent call last):
        4: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        3: from (irb):58
        2: from enumerables.rb:82:in `my_inject'
        1: from enumerables.rb:9:in `my_each'
NoMethodError (undefined method `[]' for 5..50:Range)
irb(main):059:0> 
```

File: enumerables.rb [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+    return to_enum(:my_map) unless block_given?
+
+    arr = []
+    my_each { |e| arr << block.call(e) } if block_given?
+    arr
+  end
+
+  def my_inject(initial = nil, sym = nil, &block)
+    acc = initial || first
+    acc = first if initial.class == Symbol
```
Comment (samgaco) [inline enumerables.rb]: :+1:

**Pull request merged**